genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/grpc/GrpcAgentHeartBeatServiceImpl.java

```diff
@@ -0,0 +1,178 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl.grpc;
+
+import com.netflix.genie.agent.execution.services.AgentHeartBeatService;
+import com.netflix.genie.proto.AgentHeartBeat;
+import com.netflix.genie.proto.HeartBeatServiceGrpc;
+import com.netflix.genie.proto.ServerHeartBeat;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.annotation.Validated;
+
+import javax.validation.constraints.NotBlank;
+import java.time.Instant;
+import java.util.concurrent.ScheduledFuture;
+
+/**
+ * gRPC implementation of AgentHeartBeatService.
+ * Sends heartbeats to the server.
+ * Transparently handles disconnections and stream errors by establishing a new stream.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+@Validated
+class GrpcAgentHeartBeatServiceImpl implements AgentHeartBeatService {
+
+    private static final long HEART_BEAT_PERIOD_MILLIS = 2_000; //TODO make configurable
+    private static final int STREAM_RESET_DELAY_MILLIS = 1000; //TODO make configurable
+
+    private final HeartBeatServiceGrpc.HeartBeatServiceStub client;
+    private final ThreadPoolTaskScheduler taskScheduler;
+
+    private boolean isConnected;
+    private StreamObserver<AgentHeartBeat> requestObserver;
+    private ScheduledFuture<?> heartbeatFuture;
+    private String claimedJobId;
+    private AgentHeartBeat heartBeatMessage;
+
+    GrpcAgentHeartBeatServiceImpl(
+        final HeartBeatServiceGrpc.HeartBeatServiceStub client
+    ) {
+        this.client = client;
+        this.taskScheduler = new ThreadPoolTaskScheduler();
+        this.taskScheduler.setPoolSize(1);
+        this.taskScheduler.initialize();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void start(@NotBlank final String jobId) {
+        if (StringUtils.isNotBlank(this.claimedJobId)) {
```

Comment:
In practice, this `if` statement will be invoked only once during execution, this check is not even needed in theory.
So we don't save much in terms of speed, and it would introduce an extra variable, that's why i went with the string.

Reply:
(Once these become `Lifecycle` objects, i think the "already started" check is part of the provided behavior and we can probably get rid of this completely